### ISSUE TITLE
Should support exporting annotations on annotations

### ISSUE THREAD

**Issue by rsgoncalves** [protegeproject/csv-export-plugin#1]
with a checkbox trigger in the export options to enable/disable.

**Comment by tudorache** [comment]
We need some clarification on how the export should be formatted. Maybe @fragosog could also provide feedback, too.

Consider the example in the screenshot (adapted from NCIT):

![screen shot 2016-09-27 at 3 05 41 pm](https://cloud.githubusercontent.com/assets/2709502/18893775/1c2cffe0-84c4-11e6-9973-702cb604c545.png)

Two questions:
**(1) How should the annotations on annotations be specified in the export UI?**

Currently, we specify "definition" as an export property. How do we specify that we want the def-source of definition (but maybe not def-timestamp)? Any recommendations for the export UI?

**(2) How should the annotations on annotations be exported?** 

E.g. as extra columns? 

If yes, would we have a column for each annotation on annotation? E.g.

entity | definition | def-source | def-timestamp | prop2 | prop 3 | ...

However, this format might end up with annotations on annotations that do not make any sense for the exported property value. For example, if one would also export the FULL_SYN (which has other annotations on annotations), the format would look like:

entity | definition | def-source | def-timestamp | full_syn | term-name| term-group | term-source| prop 3 | ...

(term-name, term-group and term-source do not make sense for definitions)

Basically, the problem is how to handle nested structures in a table/spreadsheet format.

Also, how do we handle multiple values for annotations on annotations? E.g. we might have two "rdfs:comment"-s on the definition of an entity.

Also: in the Protege UI, we can specify only one level of annotations on annotations. However, the **OWL spec allows nested annotations** down to any level (e.g., we might have an annotation-on-annotation-on-annotation). How do we deal with this generically for the export?

**Comment by fragosog** [comment]
On 1, at best we should be able to select them, at worst all the annotations on annotations should be exported by default.  If selectable, an annotation on annotation could wind up being a valid standalone annotation, so we should be able to distinguish between these and the ones that are associated with another entity such as on an annotation or on a restriction on a class.  Perhaps a "+" icon next to a property name could open a second widget where the user can select the annotation properties that they'd like to export if they are found annotating the primary property.  A wizard should also do the trick, but the "+" seems to fit better with Protege's editing/annotation philosophy.

On 2, all the properties of a given (sub) type should appear in the same field, if e.g. a concept has multiple "synonym" properties they should all be exported in one field, and the annotations on those synonyms should appear within the same sub field.  For instance, if we are exporting three fields, say the code, the rdfs:label, and the synonym(s), we'd have

```
code <field-delimiter> rdfs:label <field-delimiter> synonym
```

within the "synonym" field we'd have a second delimiter, say sub-delimiter, to distinguish between multiple synonyms.  In the example above perhaps for language variants of a simple case like the rdfs:label

```
<field-delimiter>     rdfs:label@en <sub-delimiter> rdfs:label@fr     <field-delimiter>
```

while for an annotated synonym containing e.g. a term-type (with rdfs:label "term type"), and a term-source (rdfs:label "term source"), we could do several things, perhaps an xml type of representation, or perhaps as name:value pairs.  Within each sub-field we'd also have to be able to distinguish between the property value and its annotations, so we'd need to enclose or use yet another sub-sub-delimiter (I'd favor the former).  For some examples:

```
<field-delimiter>  synonym1 { annotProp1:"annot value 1"; annotProp2:"annot value 2" } <sub-delimiter> synonym2 { more name-value pairs } <field-delimiter> . . . 
```

(using braces to enclose the annotations on annotations), or more xml-like

```
<field-delimiter>  synonym1 { <annotProp1>annot value 1</annotProp1><annotProp2>annot value 2</annotProp2>} <sub-delimiter> synonym2 { more xml-ish content } <field-delimiter> . . . 
```

the various delimiters selected, e.g. commas or tabs for field-delimiter, pipes, at, dollar, for sub-delimiters, colons, semi-colons, < , > used in the annotations, so on, should be escaped when found within an annotation accordingly.

The question of how to deal with deeply nested annotations on annotations:  I kind of suspected that they were allowed but Protege didn't deal with them.  I'm not sure at this time how much of an edge case this is as I haven't seen them (yet) in the wild, but without tool support they might not be used anyway.  In the case that they get used and Protege eventually supports them, my impression is that the name:value representation could be made more json-like?  Perhaps planning for the future, but my impression is that deeper nesting might not be used.

**Comment by fragosog** [comment]
ah, just saw the checkbox trigger suggestion.  Yes, that's what i was thinking with the "+" suggestion.

**Comment by rsgoncalves** [comment]
Do you think you'll want to export some (but not all) annotations on a particular annotation? Or is it generally the case that you just want all? In the latter case, we could implement a simple checkbox/trigger. In the former, I can do the extra button on each property as you said.

The single-field/cell solution is imo the best way to go. Round-tripping will be easier. It won't be so straightforward to take these exports and insert into a database. Is this something you (might) do? I favor JSON over XML syntax, easier on the eye.
